<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>GNA5012 - Applied Bioinformatics - 1&nbsp; Week 1 Workshop A - Introduction to the shell</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./W1_2.html" rel="next">
<link href="./index.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Week 1 Workshop A - Introduction to the shell</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">GNA5012 - Applied Bioinformatics</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W1_1.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Week 1 Workshop A - Introduction to the shell</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W1_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Week 1 Workshop B - Introduction shell scripting and pipelines</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W2_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Week 2 Workshop A - Bioinformatic File Types</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content column-page-right" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Week 1 Workshop A - Introduction to the shell</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>
<nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives"><span class="toc-section-number">1.1</span>  Learning Objectives</a></li>
  <li><a href="#working-on-the-command-line" id="toc-working-on-the-command-line"><span class="toc-section-number">1.2</span>  Working on the command line</a></li>
  <li><a href="#connecting-to-your-virtual-machine" id="toc-connecting-to-your-virtual-machine"><span class="toc-section-number">1.3</span>  Connecting to your Virtual Machine</a></li>
  <li><a href="#navigating-on-the-command-line" id="toc-navigating-on-the-command-line"><span class="toc-section-number">1.4</span>  Navigating on the command line</a></li>
  <li><a href="#creating-files-and-directories" id="toc-creating-files-and-directories"><span class="toc-section-number">1.5</span>  Creating files and directories</a></li>
  <li><a href="#copying-moving-and-deleting-files" id="toc-copying-moving-and-deleting-files"><span class="toc-section-number">1.6</span>  Copying, moving and deleting files</a></li>
  <li><a href="#wild-cards" id="toc-wild-cards"><span class="toc-section-number">1.7</span>  Wild cards</a></li>
  <li><a href="#viewing-file-content" id="toc-viewing-file-content"><span class="toc-section-number">1.8</span>  Viewing file content</a></li>
  <li><a href="#searching-file-content" id="toc-searching-file-content"><span class="toc-section-number">1.9</span>  Searching file content</a></li>
  <li><a href="#editing-files---sed" id="toc-editing-files---sed"><span class="toc-section-number">1.10</span>  Editing files - Sed</a></li>
  <li><a href="#editing-files---nano" id="toc-editing-files---nano"><span class="toc-section-number">1.11</span>  Editing files - Nano</a></li>
  </ul>
</nav>
<section id="learning-objectives" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="learning-objectives"><span class="header-section-number">1.1</span> Learning Objectives</h2>
<p>At the conclusion of today’s workshop students are expected to be able to:</p>
<ul>
<li><p>To compare and contrast command line interfaces with graphical user interfaces</p></li>
<li><p>To articulate the principles of Bash programming, including the common syntax of commands</p></li>
<li><p>To execute common commands to solve routine problems using Bash</p></li>
</ul>
<div class="cell">

</div>
</section>
<section id="working-on-the-command-line" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="working-on-the-command-line"><span class="header-section-number">1.2</span> Working on the command line</h2>
<p>Bioinformatic analysis is conducted using a command-line interface (CLI). This is a text-based interface that allows users to interact directly with the computers operating system by executing commands. This is in contrast with the graphical user interface (GUI) that you might be familiar with, where instructions are sent to the computer via a point-aim-click mechanism. Below is an example of how one might view the contents of a directory using a CLI (Left) and a GUI (Right). The most common shells are Bash and Zsh. These shells typically operate in Unix based operating systems. We will be using Ubuntu-based virtual machines in this course.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot 2023-07-13 at 11.54.05 am.png" class="img-fluid figure-img"></p>
</figure>
</div>
<p>Shell scripting can be very useful in bioinformatics including</p>
<ul>
<li><p>Reproducibility - Shell scripts can be saved and re-executed a later date. Commands executed in the shell are also saved and can be referred to at a later date.</p></li>
<li><p>Throughput - Many tasks in bioinformatics are repetitive. For example, if we were conducting a sequencing experiment on 100 samples and would like to trim adapters off reads we can use loops to perform this task on all sets of reads. This is much quicker than using a GUI to perform the same task.</p></li>
<li><p>Integration - Shell scripting allows you to integrate several programs in workflows. For example, might make a shell script that calls FastQC to inspect reads and then TrimGalore! to remove adapters. This can be done in a single script</p></li>
<li><p>Efficiency - Graphical user interfaces can be resource intensive. Using the shell frees resources that would usually be used for the GUI.</p></li>
</ul>
</section>
<section id="connecting-to-your-virtual-machine" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="connecting-to-your-virtual-machine"><span class="header-section-number">1.3</span> Connecting to your Virtual Machine</h2>
<p>There are several important steps to connect to your virtual machine.</p>
<p><strong>Mac OSX/Linux based computers</strong></p>
<ol type="1">
<li><p>Ensure that you are connected to the Monash VPN. Instructions for connecting to the Monash VPN can be found <a href="https://www.monash.edu/esolutions/network/vpn">here</a>.</p></li>
<li><p>Go to the search bar and type in “terminal”. This will open your machines terminal.</p></li>
<li><p>To connect to your virtual machine you need to execute the following command, replacing username with your trainee username and hostname with the corresponding hostname in the <a href="https://docs.google.com/spreadsheets/d/1H9k6jPfAR96m_0iKUA_sbLZEUzDmlME8yHlfLmAKTow/edit#gid=0">google sheet</a>:</p></li>
</ol>
<p><code>ssh username@hostname</code></p>
<ol start="4" type="1">
<li><p>You will be prompted to enter your password.</p></li>
<li><p>You may be prompted with the following message:</p></li>
</ol>
<p><code>Are you sure you want to continue connecting (yes/no/[fingerprint])?</code></p>
<p>Type ‘Yes’</p>
<ol start="6" type="1">
<li><p>You are now successfully connected to the VPN. You should see something similar to this:</p>
<p><img src="images/Screenshot 2023-07-17 at 1.24.58 pm.png" class="img-fluid"></p></li>
</ol>
<p><strong>Windows based computers</strong></p>
<ol type="1">
<li><p>Ensure that you are connected to the Monash VPN. Instructions for connecting to the Monash VPN can be found <a href="https://www.monash.edu/esolutions/network/vpn">here</a>.</p></li>
<li><p><a href="https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html">Download and install PuTTY</a>. PuTTY is an open-source SSH client for Windows and will allow you to connect to your virtual machine.</p></li>
<li><p>When you open PuTTY you will be presented with the following:</p>
<p><img src="images/Unknown-7.png" class="img-fluid"></p></li>
<li><p>In the “hostname” field, enter the hostname corresponding to your virtual machine as per the <a href="https://docs.google.com/spreadsheets/d/1H9k6jPfAR96m_0iKUA_sbLZEUzDmlME8yHlfLmAKTow/edit#gid=0">google sheet</a>. Click the “open” button.</p></li>
<li><p>You will be prompted to enter a username. Enter the username that corresponds to your virtual machine. Hit enter.</p></li>
<li><p>You will be prompted to enter a password. Enter the password that corresponds to your virtual machine. Note: Your password will not be visible as you type it. This is normal. Hit Enter.</p></li>
<li><p>You should now be connected to your virtual machine.</p></li>
</ol>
<p><img src="images/Screenshot 2023-07-17 at 1.58.11 pm.png" class="img-fluid"></p>
</section>
<section id="navigating-on-the-command-line" class="level2" data-number="1.4">
<h2 data-number="1.4" class="anchored" data-anchor-id="navigating-on-the-command-line"><span class="header-section-number">1.4</span> Navigating on the command line</h2>
<p>Navigating on the command line is an important skill and mastery of navigation will substantially reduce your time troubleshooting basic path related errors. Before we introduce the basic commands used for navigation it is important to introduce the concept of relative and absolute paths. Both can be used to specify the location of a directory or file but there is an important distinction between them.</p>
<p><strong>Absolute Path</strong>: The absolute path of a file or directory begins at the root directory and specifies the full location of a file in the file system. This is the exact path from the root <code>/</code> to the file, and does not rely on the current working directory. Absolute pathways always begin with a forward slash <code>/</code>. Here is an example of an absolute path:</p>
<ul>
<li><code>/home/trainee32/GNA5012/week_1</code></li>
</ul>
<p><strong>Relative path:</strong> The relative path of a file begins at the current working directory and tracks the path from this directory to a given file. Relative paths do not begin with <code>/</code>. Image we are currently the <code>home</code> directory. The relative path to <code>shell-lesson-data</code> would be:</p>
<ul>
<li><code>trainee32/GNA5012/week_1</code></li>
</ul>
<p><strong>When things go wrong:</strong> By far the most common error we see when navigating or pointing to a file occurs when the path to that file is set incorrectly. If this is the case you will see the following error: <code>No such file or directory</code>. If you see this error you need to double check the accuracy of the path you have provided.</p>
<p>There are three key commands that we use for navigation.</p>
<ol type="1">
<li><code>pwd</code> - This command will tell you where you are currently working.</li>
</ol>
<blockquote class="blockquote">
<p><strong>Exercise 1:</strong></p>
<p><strong>Use the <code>pwd</code> command to print out your current working directory to the terminal. What do you notice about the path that is returned? Is it absolute or relative? How can you tell?</strong></p>
</blockquote>
<ol start="2" type="1">
<li><code>cd</code> - This command will change your current working directory (hence the name cd). The syntax for this command is <code>cd desired_path</code>. <code>desired_path</code> can be absolute or relative. You will use this command to move around within your virtual environment.</li>
</ol>
<p>Let’s imagine we are in our home directory, which in this case is <code>/users/student1</code></p>
<table class="table">
<colgroup>
<col style="width: 15%">
<col style="width: 7%">
<col style="width: 13%">
<col style="width: 62%">
</colgroup>
<thead>
<tr class="header">
<th>Current working directory</th>
<th>Command</th>
<th>New working directory</th>
<th>Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>/users/student1</code></td>
<td><code>cd .</code></td>
<td><code>/users/student1</code></td>
<td>The <code>.</code> operator stands for your present directory. <code>cd</code> will not alter the working directory in response to this</td>
</tr>
<tr class="even">
<td><code>/users/student1</code></td>
<td><code>cd ..</code></td>
<td><code>/users</code></td>
<td>The <code>..</code> operator will move you one level up the tree to the parent directory.</td>
</tr>
<tr class="odd">
<td><code>/users/student1</code></td>
<td><code>cd week_1</code></td>
<td><code>/users/student/week_1</code></td>
<td>This command will move you to the <code>week_1</code> directory which resides in your working directory.</td>
</tr>
<tr class="even">
<td><code>/users/student/week_1</code></td>
<td><code>cd ~</code></td>
<td><code>/users/student1</code></td>
<td>The <code>~</code> symbol represents your home directory and we are thus returned there</td>
</tr>
<tr class="odd">
<td><code>/users/student1</code></td>
<td><code>cd /</code></td>
<td><code>/</code></td>
<td>The <code>/</code> symbol represents the root of the filesystem. This command will take you to the very top of the tree!</td>
</tr>
</tbody>
</table>
<blockquote class="blockquote">
<p><strong>Exercise 2:</strong></p>
<p><strong>Use the <code>cd</code> command to navigate to the week_1 directory. You can use either relative or absolute paths to achieve this. Note that if you use the absolute path, you should take note of your current working directory as noted in exercise 1.</strong></p>
</blockquote>
<ol start="3" type="1">
<li><code>ls</code> - The last command in this series is the ls command. This command with list the contents of your current working directory. This is akin to viewing the contents of a directory using a graphical user interface.</li>
</ol>
<blockquote class="blockquote">
<p><strong>Exercise 3:</strong></p>
<p><strong>Use the <code>ls</code> command to list the contents of the week_1 directory. There are several options associated with this command. You can view these (and options associated with any other command) using the man command. Try out a few, especially the -a option. What do you notice?</strong></p>
</blockquote>
</section>
<section id="creating-files-and-directories" class="level2" data-number="1.5">
<h2 data-number="1.5" class="anchored" data-anchor-id="creating-files-and-directories"><span class="header-section-number">1.5</span> Creating files and directories</h2>
<p>We can create files and directories using the <code>touch</code> and <code>mkdir</code> commands. Touch will create an empty file and mkdir will create a new directory. Lets give it a try:</p>
<blockquote class="blockquote">
<p><strong>Exercise 4:</strong></p>
<p><strong>Within the</strong> <code>~/GNA5012/week_1</code> <strong>directory, create a directory called “src”. Within this directory, create an empty file called script_1.sh</strong></p>
</blockquote>
</section>
<section id="copying-moving-and-deleting-files" class="level2" data-number="1.6">
<h2 data-number="1.6" class="anchored" data-anchor-id="copying-moving-and-deleting-files"><span class="header-section-number">1.6</span> Copying, moving and deleting files</h2>
<p>There are occasions when we want to copy, move or even delete files from the shell. This can be achieved using the <code>cp</code>, <code>mv</code> and <code>rm</code> functions, respectively.</p>
<ul>
<li>The basic syntax for <code>cp</code> and <code>mv</code> is as follows: <code>cp source_file target_file</code></li>
</ul>
<blockquote class="blockquote">
<p><strong>Exercise 5:</strong></p>
<p><strong>Within the `</strong><code>~/GNA5012/week_1`</code> <strong>directory, create a couple of files using the touch command. Explore the <code>cp</code> and <code>mv</code> commands by moving these files around and creating new copies. When you are done clean up the files you have created with <code>rm.</code></strong></p>
</blockquote>
</section>
<section id="wild-cards" class="level2" data-number="1.7">
<h2 data-number="1.7" class="anchored" data-anchor-id="wild-cards"><span class="header-section-number">1.7</span> Wild cards</h2>
<p>Sometimes when we are performing tasks in bash we would like to identify patterns in text strings. For instance, I might have a large set of fastq files from a paired-end sequencing run. Read pairs are usually denoted by R1 and R2. I can use wild cards to identify all the R1 fastq files in a directory in a directory by using the following command</p>
<p><code>ls *R1.fastq</code></p>
<p>There are several wildcards in Bash:</p>
<table class="table">
<caption>These can be integrated into any commands where finding strings and patterns are useful. Several wildcards can be combined within a single string, making them a very powerful tool.</caption>
<colgroup>
<col style="width: 4%">
<col style="width: 24%">
<col style="width: 10%">
<col style="width: 59%">
</colgroup>
<thead>
<tr class="header">
<th>Wildcard</th>
<th>Description</th>
<th>Example</th>
<th>Output</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>*</td>
<td>Matches any number of characters.</td>
<td><code>ls *.pdf</code></td>
<td>All files with the <code>.pdf</code> file extension</td>
</tr>
<tr class="even">
<td>?</td>
<td>Matches any single character.</td>
<td><code>ls sample_1.?am</code></td>
<td>All files with an extension that starts with any character and is followed by <code>am</code>. Examples might include <code>sample_1.bam</code> or <code>sample_1.sam</code></td>
</tr>
<tr class="odd">
<td>[]</td>
<td>Matches any single character within a range or set</td>
<td><code>ls sample_[123].sam</code></td>
<td>Returns <code>sample_1.sam</code>, <code>sample_2.sam</code>, and <code>sample_3.sam</code> if they exist</td>
</tr>
<tr class="even">
<td>{}</td>
<td>Expands a comma-separated list of strings or characters</td>
<td><code>ls sample_1.{sam,bam}</code></td>
<td>Returns <code>sample_1.sam</code> and <code>sample_1.bam</code></td>
</tr>
</tbody>
</table>
<blockquote class="blockquote">
<p><strong>Exercise 6:</strong></p>
<p><strong>Let’s try a few examples. In the</strong> <code>~/GNA5012/week_1/data/sequences</code> <strong>you will find a set of DNA sequences in FASTA format. Each file contains a different number of sequences. FASTA files contain a descriptor line, which is indicated by the <code>&gt;</code></strong> <strong>symbol, and a line containing the sequence of nucleotides. We can use the <code>wc</code> tool to evaluate the number of lines in a file. Use the <code>wc</code></strong> <strong>tool and wild-card expressions to determine the number of sequences in each file. (Hint - if you are unsure about <code>wc</code></strong>, <strong>check out the manual using <code>man wc</code></strong></p>
</blockquote>
<p><strong>Solution</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> <span class="pp">*</span>.fasta <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="viewing-file-content" class="level2" data-number="1.8">
<h2 data-number="1.8" class="anchored" data-anchor-id="viewing-file-content"><span class="header-section-number">1.8</span> Viewing file content</h2>
<p>There are several ways we can work with the contents of files in Bash. To view files we can leverage a suite of in-built tools, including <code>cat</code>, <code>less</code>, <code>head</code> and <code>tail</code></p>
<table class="table">
<colgroup>
<col style="width: 5%">
<col style="width: 73%">
<col style="width: 21%">
</colgroup>
<thead>
<tr class="header">
<th>Command</th>
<th>Description</th>
<th>Use case</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>cat</code></td>
<td>This command will concatenate files and return the contents to standard output (the terminal).</td>
<td><ul>
<li><p>To view the entire contents of a file</p></li>
<li><p>To combine several files</p></li>
</ul></td>
</tr>
<tr class="even">
<td><code>less</code></td>
<td>This command will display the contents of a file in your terminal window. You can move through the file using the up and down arrows. You exit the program by typing <code>q</code>.</td>
<td><ul>
<li>To view and navigate the contents of a file</li>
</ul></td>
</tr>
<tr class="odd">
<td><code>head</code></td>
<td>This command will print out the first n lines of a file. By default this is 10. Different options can be set to modify this.</td>
<td><ul>
<li>View the header of a file</li>
</ul></td>
</tr>
<tr class="even">
<td><code>tail</code></td>
<td>This command will print out the last n lines of a file. By default this is 10. Different options can be set to modify this.</td>
<td><ul>
<li>View the end of a file</li>
</ul></td>
</tr>
</tbody>
</table>
<blockquote class="blockquote">
<p><strong>Exercise 7:</strong></p>
<p><strong>In the</strong> <code>~/GNA5012/week_1/data/vcf</code> <strong>you will find a large variant call file (we will learn what the contents of this file mean in later weeks). Using the commands above, explore this file. What do you think the most effective way to navigate this file?</strong></p>
</blockquote>
<p><strong>Solution</strong></p>
<p>My typical approach here is to use <code>head</code> and <code>tail</code> first to inspect the top and bottom of files and then to use <code>less</code> to navigate them more easily</p>
</section>
<section id="searching-file-content" class="level2" data-number="1.9">
<h2 data-number="1.9" class="anchored" data-anchor-id="searching-file-content"><span class="header-section-number">1.9</span> Searching file content</h2>
<p>Sometimes we will want to search and extract certain phrases from files. For this we can use the <code>grep</code> tool. <code>grep</code> will search for patterns in a file and return them to the terminal. The <code>grep</code> tool has several handy options including regular pattern matching, inverted search, and pattern counting. This is an important tool that you will use in several workflows, check out the manual to see all the relevant options.</p>
<blockquote class="blockquote">
<p><strong>Exercise 8:</strong></p>
<p><strong>Earlier we counted the number of sequences in each fasta file in</strong> <code>~/GNA5012/week_1/data/sequence</code><strong>. Now, using <code>grep</code> and the wildcards that you learned earlier:</strong></p>
<ul>
<li><p><strong>Search for all sequences in the fasta files containing the string <code>TTG</code>.</strong></p></li>
<li><p><strong>Search for all sequences that don’t contain <code>TTG</code></strong></p></li>
<li><p><strong>Search for all sequences that contain <code>AAA</code> in <code>sequence_1.fasta</code> to <code>sequence_5.fasta</code></strong></p></li>
</ul>
</blockquote>
<p><strong>Solution</strong></p>
<ul>
<li><p><code>grep 'TTG' *.fasta</code></p></li>
<li><p><code>grep -v 'TTG' *.fasta</code></p></li>
<li><p><code>grep 'AAA' sequence_[15].fasta</code></p></li>
</ul>
</section>
<section id="editing-files---sed" class="level2" data-number="1.10">
<h2 data-number="1.10" class="anchored" data-anchor-id="editing-files---sed"><span class="header-section-number">1.10</span> Editing files - Sed</h2>
<p>There are two main ways we can edit files:</p>
<ul>
<li><p>Interactively using tools like <code>nano</code> and <code>vim</code></p></li>
<li><p>Non-interactively using tools like <code>sed</code></p></li>
</ul>
<p>Lets talk about <code>sed</code> first. This tool is most useful when incorporated into scripts as it can search for patterns and perform operations such as replacing the string, deleting the string and inserting characters before and after the string. The power in the tool lies in its ability to perform these operations at speed and scale. One of the most popular implementations of the <code>sed</code> tool is the substitute program. It follows the following syntax:</p>
<ul>
<li><code>sed [OPTIONS] 'command(s)' input_file(s)</code></li>
</ul>
<p>The most common applications of <code>sed</code> is the substitute function</p>
<ul>
<li>Substitute - This command can be used to replace patterns with new strings. The syntax for <code>sed</code> substitute is <code>sed 's/old_text/new_text/g' input_file</code></li>
</ul>
<blockquote class="blockquote">
<p><strong>Exercise 9: Some genome reference builds contain the <code>chr</code> prefix to chromosomes, while others omit this information. The <code>vcf</code> file we have provided you contains the <code>chr</code> prefix. Using <code>sed</code> and other commands perform the following tasks:</strong></p>
<ul>
<li><p><strong>Identify the number of variants in this file (Hint: You might consider using the <code>grep</code> and <code>wc</code> tools.</strong></p></li>
<li><p><strong>Using the <code>sed</code> tool, remove the <code>chr</code> prefix from both the header and the variant entries.</strong></p></li>
</ul>
</blockquote>
<p><strong>Solution</strong></p>
<ul>
<li><p><code>grep -v '#' HG002.vcf | wc -l</code></p></li>
<li><p><code>sed 's/chr//g' HG002.vcf</code></p></li>
</ul>
</section>
<section id="editing-files---nano" class="level2" data-number="1.11">
<h2 data-number="1.11" class="anchored" data-anchor-id="editing-files---nano"><span class="header-section-number">1.11</span> Editing files - Nano</h2>
<p>We can also edit files interactively, much like you’d use word or notepad on a GUI-based interface. To get started, watch this short video on how to use nano:</p>
<div class="quarto-video ratio ratio-16x9"><iframe src="https://www.youtube.com/embed/cLyUZAabf40" title="" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen=""></iframe></div>
<blockquote class="blockquote">
<p><strong>Exercise 10:</strong></p>
<p><strong>Let’s construct our first bash script using nano. Remember earlier we created a file called <code>script_1.sh</code></strong> <strong>in the src directory. Using nano:</strong></p>
<ul>
<li><p><strong>Open the file</strong></p></li>
<li><p><strong>Type <code>#!/bin/bash</code>, this is what we call a shebang. This provides a path to the bash interpreter and should be at the top of all of your scripts.</strong></p></li>
<li><p><strong>On a new line, write a line of code that uses <code>echo</code></strong> <strong>tool to print out “This is my first script in GNA5012”. If you are not familar with this tool, use the <code>man</code></strong> <strong>tool to review the manual.</strong></p></li>
<li><p><strong>Save the script and execute it by typing <code>bash script_1.sh</code></strong> <strong>into the terminal. If you get a path error, you may need to use <code>cd</code></strong> <strong>to ensure you are in the right directory</strong></p></li>
</ul>
</blockquote>
<p><strong>Solution:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'This is my first script in GNA5012'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p><strong>Exercise 11:</strong></p>
<p><strong>Let’s now circle back to <code>sed</code>. Using <code>sed</code> substitute:</strong></p>
<ul>
<li><p><strong>Replace GNA5012 with Applied Bioinformatics</strong></p></li>
<li><p><strong>Use the <code>&gt;</code> operator to save the output to a new file called <code>script_2.sh</code></strong></p></li>
<li><p><strong>Execute the new script.</strong></p></li>
</ul>
</blockquote>
<p><strong>Solution</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">'s/GNA5012/Applied Bioinformatics/g'</span> script_1.sh <span class="op">&gt;</span> script_2.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation column-page-right">
  <div class="nav-page nav-page-previous">
      <a href="./index.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Preface</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./W1_2.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Week 1 Workshop B - Introduction shell scripting and pipelines</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>