<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>GNA5012 - Applied Bioinformatics - 5&nbsp; Week 3 Workshop A - Quality control and read trimming</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./W3_2.html" rel="next">
<link href="./W2_2.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Week 3 Workshop A - Quality control and read trimming</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">GNA5012 - Applied Bioinformatics</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W1_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Week 1 Workshop A - Introduction to the shell</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W1_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Week 1 Workshop B - Introduction shell scripting and pipelines</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W2_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Week 2 Workshop A - Bioinformatic File Types</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W2_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Week 2 Workshop B - Quality control inspection of sequencing reads</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W3_1.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Week 3 Workshop A - Quality control and read trimming</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W3_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Week 3 Workshop B - Sequence Alignment</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content column-page-right" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Week 3 Workshop A - Quality control and read trimming</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>
<nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives"><span class="toc-section-number">5.1</span>  Learning Objectives</a></li>
  <li><a href="#workshop-setup" id="toc-workshop-setup"><span class="toc-section-number">5.2</span>  Workshop setup</a></li>
  <li><a href="#performing-quality-trimming-with-trimgalore" id="toc-performing-quality-trimming-with-trimgalore"><span class="toc-section-number">5.3</span>  Performing quality trimming with TrimGalore!</a></li>
  <li><a href="#interpreting-the-results-of-trimgalore" id="toc-interpreting-the-results-of-trimgalore"><span class="toc-section-number">5.4</span>  Interpreting the results of TrimGalore!</a></li>
  <li><a href="#performing-trimgalore-on-several-read-sets-using-loops" id="toc-performing-trimgalore-on-several-read-sets-using-loops"><span class="toc-section-number">5.5</span>  Performing TrimGalore! on several read sets using loops</a></li>
  </ul>
</nav>
<p>Last week we used the <code>FastQC</code> tool to evaluate the quality of an exome sequencing experiment performed as part of the Genome In A Bottle project. The report generated by <code>FastQC</code> identified several quality concerns that could impact downstream analyses. This is common with sequencing data, and it is exceedingly rare that, even with the best sequencing runs, you will not encounter some quality issues. As bioinformaticians, one of our main tasks is to identify and rectify these issues where possible. Once we receive the data we can perform a couple of bioinformatic preprocessing steps to try and fix these concerns, however not all quality issues can be resolved computationally. The adage of ‘Garbage in, Garbage out’ is often true - this is why we like to get involved in experiments as soon as possible, and before data is generated! In this workshop, we will use a popular read trimming tool (<code>TrimGalore!</code>) to remove sequencing adapters and perform basic sequence quality correction.</p>
<div class="cell">

</div>
<section id="learning-objectives" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="learning-objectives"><span class="header-section-number">5.1</span> Learning Objectives</h2>
<p>At the conclusion of today’s workshop students are expected to be able to:</p>
<ul>
<li><p>Setup project directories and architectures using best practices for organising computational analyses</p></li>
<li><p>Perform quality control and read trimming on exome sequencing data</p></li>
<li><p>Write a while read loop to perform quality control trimming on a large sequencing experiment corresponding to many samples</p></li>
<li><p>Compare and contrast quality metrics before and after read trimming</p></li>
</ul>
</section>
<section id="workshop-setup" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="workshop-setup"><span class="header-section-number">5.2</span> Workshop setup</h2>
<p>There is no specific setup for today’s workshop. In the first 15 minutes we will discuss how to organise projects effectively. This will include a short presentation and setting up a mock project in by following in real time. We will cover things like consistent project architecture and directory structure for reproducible and organised bioinformatic analysis. Setting projects up correctly will help you keep track of your analysis and be important over the coming weeks as you perform several different analyses, each generated a number of output files.</p>
</section>
<section id="performing-quality-trimming-with-trimgalore" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="performing-quality-trimming-with-trimgalore"><span class="header-section-number">5.3</span> Performing quality trimming with TrimGalore!</h2>
<p><a href="https://github.com/FelixKrueger/TrimGalore/blob/master/Docs/Trim_Galore_User_Guide.md">TrimGalore!</a> is a tool that wraps <a href="https://cutadapt.readthedocs.io/en/stable/">cutadapt</a> and <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">fastqc</a> into a convenient, flexible and easy to use tool. When correctly configured, TrimGalore! can:</p>
<ul>
<li><p>Perform basic quality trimming using the cutadapt method (with both default and user-set phred thresholds)</p></li>
<li><p>Remove adapter sequences</p></li>
<li><p>Perform hard trimming (removing n number of bases from either 3’ or 5’ end of the reads)</p></li>
<li><p>Remove reads below a threshold length</p></li>
<li><p>Handle paired-end sequences in a manner that ensures only reads that are paired are written to a file</p></li>
<li><p>Run <code>Fastqc</code> analysis on trimmed reads</p></li>
</ul>
<p>Lets explore the default behaviour of TrimGalore! This tool will trim based from the 3’ end of a file (as per the BWA and cutadapt method) until it reaches a nucleotide that exceeds the quality threshold. The phred threshold is Q20, which corresponds to a 1% probability of an incorrect basecall. This can be modified using the <code>-q</code> option. For most analyses, Q20 represents a good trade off between quality and lost reads. TrimGalore! will also simultaneously detect and remove adapter sequences. TrimGalore! will automatically detect the following adapters:</p>
<ul>
<li><p>Illumina standard adapters: <code>AGATCGGAAGAGC</code></p></li>
<li><p>Illumina small RNA adapters: <code>TGAATTCTCGG</code></p></li>
<li><p>Nextera adapters: <code>CTGTCTTATA</code></p></li>
</ul>
<p>We can also specify the adapter sequence if we are using non-standard adapters using the <code>-a</code> option. Other default parameters are as follows:</p>
<table class="table">
<colgroup>
<col style="width: 14%">
<col style="width: 9%">
<col style="width: 76%">
</colgroup>
<thead>
<tr class="header">
<th>Parameter and Option Flag</th>
<th>Default</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>--stringency</td>
<td>1</td>
<td>Overlap with adapter sequence required to trigger trimming.</td>
</tr>
<tr class="even">
<td>--length</td>
<td>20</td>
<td>Minimum length required to retain a read after trimming. Paired-end requires both sequence meet this threshold.</td>
</tr>
<tr class="odd">
<td>-j</td>
<td>1</td>
<td>Number of cores used in analysed. Required python3 and pigz packages.</td>
</tr>
<tr class="even">
<td>--paired</td>
<td>Single-end reads</td>
<td>Whether the reads are from a paired-end or single-end experiment. Note that <strong>both</strong> reads must be provided as input if this option is selected</td>
</tr>
</tbody>
</table>
<p>For the purposes of this workshop and your assessment to leave the majority of the default parameters in place. Lets move onto the standard syntax of TrimGalore!</p>
<p>Here is an example of how to write a TrimGalore script to analyse paired end sequencing:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="va">RESULTS</span><span class="op">=</span>PATH_TO_RESULTS</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="va">READ_1</span><span class="op">=</span>PATH_TO_READ_1</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="va">READ_2</span><span class="op">=</span>PATH_TO_READ_2</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">trim_galore</span> <span class="at">--fastqc</span> <span class="at">--paired</span> <span class="at">--output_dir</span> <span class="va">$RESULTS</span> <span class="va">$READ_1</span> <span class="va">$READ_2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p>First we set three variables, corresponding to the desired path of our trimmed reads results files, and the location of the forward and reverse reads.</p></li>
<li><p><code>trim_galore</code> - This is the tool we are calling.</p></li>
<li><p><code>--fastqc</code> - This instructs TrimGalore! to run <code>fastqc</code> on the sequencing reads after trimming.</p></li>
<li><p><code>--paired</code> - This informs TrimGalore! that we are conducting trimming on paired-end sequences and changes the behaviour of the tool accordingly.</p></li>
<li><p><code>--output_dir</code> - This instructs TrimGalore! where to write the trimming report, trimmed reads and the <code>fastqc</code> reports.</p></li>
</ul>
<blockquote class="blockquote">
<p><strong>Exercise 1: Trimming reads using TrimGalore!</strong></p>
<ul>
<li><p>In the <code>~/GNA5012/week2_5/src</code> directory create a new script called <code>trim_reads.sh</code></p></li>
<li><p>Using the above template, create a script that will run TrimGalore! on a set of reads in the <code>data</code> directory.</p></li>
<li><p>You may choose to include variables within your script, or use command line arguments to direct your script to the <code>fastq</code> files. Remember to set the <code>output_dir</code> appropriately to ensure the trimmed reads and associated files are written to the correct location.</p></li>
</ul>
</blockquote>
<p><strong>Solution:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co">##Usage bash trim_reads.sh read1 read2</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="va">OUTDIR</span><span class="op">=</span>~/GNA5012/week2_5/results/trimmed/</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">trim_galore</span> <span class="at">--fastqc</span> <span class="at">--paired</span> <span class="at">--output_dir</span> <span class="va">$OUTDIR</span> <span class="va">${1}</span> <span class="va">${2}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="interpreting-the-results-of-trimgalore" class="level2" data-number="5.4">
<h2 data-number="5.4" class="anchored" data-anchor-id="interpreting-the-results-of-trimgalore"><span class="header-section-number">5.4</span> Interpreting the results of TrimGalore!</h2>
<p>TrimGalore! will generate several files, including trimmed <code>fastq</code> files, and trimming and <code>fastqc</code> reports. If you have successfully ran TrimGalore! your <code>results/trimmed/</code> directory will look something like this:</p>
<p><img src="images/Screenshot 2023-08-08 at 9.47.24 pm.png" class="img-fluid"></p>
<p>Let’s explore what some of these files contain:</p>
<ul>
<li><p><code>*_val_[12].fq.gz</code>: These two files are your trimmed reads, corresponding to the trimmed forward and reverse read. These are the files that you will use in your sequence alignments.</p></li>
<li><p><code>*_trimming_report.txt:</code> These files contain important metrics related to the trimming performed by TrimGalore!</p></li>
<li><p><code>*_val_[12]_fastqc.html</code>: These are the reports from the post-trimming <code>fastqc</code> analysis. These should be compared to the <code>fastqc</code> files that you generated prior to trimming during last weeks workshop.</p></li>
</ul>
<blockquote class="blockquote">
<p><strong>Exercise 2: Evaluating TrimGalore! results</strong></p>
<ul>
<li><p>Download the FastQC reports to your computer using CyberDuck or winSCP. Compare them with the the untrimmed FastQC reports.</p>
<ul>
<li><p>What do you notice?</p></li>
<li><p>Take notes on what metrics have changed and what impact this may have on alignment and variant calling.</p></li>
</ul></li>
<li><p>View the trimming report files.</p>
<ul>
<li><p>How many reads contained adapters?</p></li>
<li><p>How many reads were filtered?</p></li>
<li><p>How many base-pairs were trimmed?</p></li>
</ul></li>
</ul>
</blockquote>
</section>
<section id="performing-trimgalore-on-several-read-sets-using-loops" class="level2" data-number="5.5">
<h2 data-number="5.5" class="anchored" data-anchor-id="performing-trimgalore-on-several-read-sets-using-loops"><span class="header-section-number">5.5</span> Performing TrimGalore! on several read sets using loops</h2>
<p>It is rare that we would perform analysis on a single sample or set of reads. We would typically analyse a sequencing run which can contain anywhere from 8 to 24+ samples. To automate this process we can employ loops. Simple <code>for</code> loops are not well suited for this type of analysis as we are dealing with paired end reads. Instead we can use <code>while read</code> loops to iterate over a sample sheet containing our sample names. There are two important elements to a <code>while read</code> loop:</p>
<ul>
<li><p>The input file: The sample sheet should contain the base name without an extensions. We can construct this using <code>ls</code> <code>grep</code> and <code>sed.</code></p></li>
<li><p>The loop structure: The loop should be structured to contain the necessary commands and call the variable that we set in addition to its extension.</p></li>
</ul>
<p>Here is an example of how we might employ a <code>while read</code> loop to perform read trimming:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co">### In this example, imagine the reads are in the following file path: </span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ~/GNA5012/week2_5/data/sequencing_run/</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co">### Also imagine that the file names are in the following standard format:</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Samp_1_R1.fastq.gz Samp_1_R2.fastq.gz ...</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># List all files in ~/GNA5012/week2_5/data/sequencing_run/ </span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Select only those containing the string _R1.fastq.gz. </span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># We only do this for reads in one direction to avoid duplicates. </span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove this string to be left with the base name.  Write to file called sample_list.txt</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> ~/GNA5012/week2_5/data/sequencing_run/ <span class="kw">|</span> <span class="fu">grep</span> <span class="st">'R1.fastq.gz'</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/_R1.fastq.gz//g'</span> <span class="op">&gt;</span> sample_list.txt</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co"># While read loop to read each line of the sample_list.txt file and assign it to the ${sample} variable</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="cf">while</span> <span class="bu">read</span> <span class="va">sample</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Set results directory</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="va">RESULTS</span><span class="op">=</span>~/GNA5012/week2_5/results/trimmed/</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="va">DATA</span><span class="op">=</span>~/GNA5012/week2_5/data/sequencing_run</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Run trim_galore in paired end mode</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="ex">trim_galore</span> <span class="at">--fastqc</span> <span class="at">--paired</span> <span class="at">--output_dir</span> <span class="va">$RESULTS</span> <span class="va">$DATA</span>/<span class="va">${sample}</span>_R1.fastq.gz <span class="va">$DATA</span>/<span class="va">${sample}</span>_R2.fastq.gz</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Provide sample_list.txt file as standard input using the &lt; redirect</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span> <span class="op">&lt;</span> sample_list.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now it’s your turn:</p>
<blockquote class="blockquote">
<p><strong>Exercise 3: Performing TrimGalore! on several read sets using loops</strong></p>
<ul>
<li><p>Create a new directory in <code>results/trimmed</code> called <code>sequencing_run</code>. This will be your results directory.</p></li>
<li><p>In the <code>src</code> directory create a new script and write a while read loop to iterate over all the samples in <code>data/sequencing_run</code> and performed TrimGalore! on them in paired-end mode.</p></li>
<li><p>Once you have successfully executed your loop, remove the <code>results/trimmed/sequencing_run</code> directory using <code>rm -rf results/trimmed/sequencing_run/</code></p></li>
</ul>
</blockquote>
<p><strong>Solution:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># List all files in ~/GNA5012/week2_5/data/sequencing_run/</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Select only those containing the string _1.fastq.gz.</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># We only do this for reads in one direction to avoid duplicates.</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove this string to be left with the base name.  Write to file called sample_list.txt</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> ~/GNA5012/week2_5/data/sequencing_run/ <span class="kw">|</span> <span class="fu">grep</span> <span class="st">'_1.fastq.gz'</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/_1.fastq.gz//g'</span> <span class="op">&gt;</span> sample_list.txt</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="cf">while</span> <span class="bu">read</span> <span class="va">sample</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Set results directory</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="va">RESULTS</span><span class="op">=</span>~/GNA5012/week2_5/results/trimmed/sequencing_run/</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="va">DATA</span><span class="op">=</span>~/GNA5012/week2_5/data/sequencing_run</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Run trim_galore in paired end mode</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="ex">trim_galore</span> <span class="at">--fastqc</span> <span class="at">--paired</span> <span class="at">--output_dir</span> <span class="va">$RESULTS</span> <span class="va">$DATA</span>/<span class="va">${sample}</span>_1.fastq.gz <span class="va">$DATA</span>/<span class="va">${sample}</span>_2.fastq.gz</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Provide sample_list.txt file as standard input using the &lt; redirect</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span> <span class="op">&lt;</span> sample_list.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation column-page-right">
  <div class="nav-page nav-page-previous">
      <a href="./W2_2.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Week 2 Workshop B - Quality control inspection of sequencing reads</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./W3_2.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Week 3 Workshop B - Sequence Alignment</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>