<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>GNA5012 - Applied Bioinformatics - 9&nbsp; Week 4 Workshop B - Variant Calling</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./W4_1.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Week 4 Workshop B - Variant Calling</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">GNA5012 - Applied Bioinformatics</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assessment.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Assessment 1 - Capstone A</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W1_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Week 1 Workshop A - Introduction to the shell</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W1_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Week 1 Workshop B - Introduction shell scripting and pipelines</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W2_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Week 2 Workshop A - Bioinformatic File Types</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W2_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Week 2 Workshop B - Quality control inspection of sequencing reads</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W3_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Week 3 Workshop A - Quality control and read trimming</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W3_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Week 3 Workshop B - Sequence Alignment</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W4_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Week 4 Workshop A - Post-Alignment Processing</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W4_2.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Week 4 Workshop B - Variant Calling</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content column-page-right" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Week 4 Workshop B - Variant Calling</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>
<nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives"><span class="toc-section-number">9.1</span>  Learning Objectives</a></li>
  <li><a href="#workshop-setup" id="toc-workshop-setup"><span class="toc-section-number">9.2</span>  Workshop Setup</a></li>
  <li><a href="#calling-variants-with-strelka2" id="toc-calling-variants-with-strelka2"><span class="toc-section-number">9.3</span>  Calling variants with Strelka2</a></li>
  <li><a href="#running-the-strelka2-workflow" id="toc-running-the-strelka2-workflow"><span class="toc-section-number">9.4</span>  Running the Strelka2 workflow</a></li>
  <li><a href="#evaluating-strelka2-variant-calling" id="toc-evaluating-strelka2-variant-calling"><span class="toc-section-number">9.5</span>  Evaluating Strelka2 Variant calling</a></li>
  <li><a href="#empirical-variant-scoring-filtration-in-strelka2" id="toc-empirical-variant-scoring-filtration-in-strelka2"><span class="toc-section-number">9.6</span>  Empirical variant scoring filtration in Strelka2</a></li>
  </ul>
</nav>
<p>This chapter marks a crucial juncture in our workshop manual, focusing on germline variant calling through Strelka2. Having completed the preprocessing phase, specifically the task of marking duplicates in our aligned BAM file, we now lay the groundwork for accurate variant calling. In this workshop , we leverage Strelka2’s potential in germline variant analysis. Strelka2 is a sophisticated software tool designed for accurate and efficient variant calling from high-throughput sequencing data. It can be used for both germline and somatic variant calling, however today we focus on its germline variant calling functions.</p>
<div class="cell">

</div>
<section id="learning-objectives" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="learning-objectives"><span class="header-section-number">9.1</span> Learning Objectives</h2>
<p>At the conclusion of today’s workshop students are expected to be able to:</p>
<ul>
<li><p>Perform germline variant calling analysis on exome sequencing data</p></li>
<li><p>Critically analyse variant calling metrics</p></li>
</ul>
</section>
<section id="workshop-setup" class="level2" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="workshop-setup"><span class="header-section-number">9.2</span> Workshop Setup</h2>
<p>Successful completion of today’s workshop is predicated on having successfully marked duplicates and indexed the resulting BAM file on Wednesday.</p>
<p>Before starting today’s workshop create two new directories using the following commands:</p>
<ul>
<li><p><code>cd GNA5012/week2_5</code></p></li>
<li><p><code>mkdir results/variant_calling</code></p></li>
<li><p><code>mkdir results/variant_calling/NIST7035</code></p></li>
</ul>
</section>
<section id="calling-variants-with-strelka2" class="level2" data-number="9.3">
<h2 data-number="9.3" class="anchored" data-anchor-id="calling-variants-with-strelka2"><span class="header-section-number">9.3</span> Calling variants with Strelka2</h2>
<p>Strelka2 utilises a tiered haplotype model that leverages both alignment and local assembly based haplotyping to call variants. A distinctive aspect of this approach is its utilises of local assembly to enhance the precision of variant calling. Instead of relying solely on the aligned reads, Strelka2 employs de novo assembly of the sequences surrounding a suspected variant site. This approach helps resolve challenging genomic regions, such as those containing insertions, deletions, or repetitive sequences.</p>
<p>Strelka2 identifies potential germline variants by comparing the local assemblies of the individual’s germline sample with the matched reference sample. Deviations between the two assemblies indicate the presence of genetic variations.</p>
<p>To execute Strelka2 we first have to configure the analysis. We do this using the <code>configureStrelkaGermlineWorkflow.py</code> python script. This configuration script will contain the path to the input BAM file, the reference genome, the output directory and any additional options that are relevant to the analysis. Here is an example configuration script:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="va">STRELKA</span><span class="op">=</span>/mnt/nfs/ubuntu/strelka-2.9.2.centos6_x86_64/bin/</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="va">ALN_DIRECTORY</span><span class="op">=</span>PATH_TO_ALIGNMENT_DIRECTORY</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="va">SAMPLE</span><span class="op">=</span>SAMPLE_NAME</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="va">OUTPUT</span><span class="op">=</span>PATH_TO_OUTPUT_DIRECTORY</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="va">REF</span><span class="op">=</span>~/GNA5012/week2_5/ref/genome.fa</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="ex">python2</span> <span class="va">${STRELKA}</span>/configureStrelkaGermlineWorkflow.py <span class="dt">\</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>--bam <span class="va">${ALN_DIRECTORY}</span>/<span class="va">${SAMPLE}</span>_marked_duplicates.bam <span class="dt">\</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>--referenceFasta <span class="va">$REF</span> <span class="dt">\</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>--runDir <span class="va">${OUTPUT}</span>/<span class="va">${SAMPLE}</span> <span class="dt">\</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>--exome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Lets break down each line.</p>
<ul>
<li><p><code>#!/bin/bash</code> - As always, we begin with a shebang to direct the script to the bash interpreter</p></li>
<li><p><code>STRELKA=/mnt/nfs/ubuntu/strelka-2.9.2.centos6_x86_64/bin/</code> - This is the absolute path to the <code>Strelka</code> binary directory containing important scripts.</p></li>
<li><p><code>ALN_DIRECTORY=PATH_TO_ALIGNMENT_DIRECTORY</code> - This variable sets our alignment directory.</p></li>
<li><p><code>SAMPLE=SAMPLE_NAME</code> - This variable corresponds to your sample name</p></li>
<li><p><code>OUTPUT=PATH_TO_OUTPUT_DIRECTORY</code> - This variable points to your output directory</p></li>
<li><p><code>REF=~/GNA5012/week2_5/ref/genome.fa</code> - This variable points to your reference genome, which in our case is GRCh38.</p></li>
<li><p><code>python2 configureStrelkaGermlineWorkflow.py \</code> - python2 is the programming language that we will use to execute the configuration script.</p></li>
<li><p><code>--bam ${ALN_DIRECTORY}/${SAMPLE}_marked_duplicates.bam \</code> - This option provides our input BAM file.</p></li>
<li><p><code>--referenceFasta $REF \</code> - Option pointing to the reference genome</p></li>
<li><p><code>--runDir ${OUTPUT}/${SAMPLE} \</code> - Option pointing to our output directory</p></li>
<li><p><code>--exome</code> - Option indicating that we are analysing exome data.</p></li>
</ul>
<blockquote class="blockquote">
<p><strong>Exercise 1: Write a script to configure Strelka2 to call variants on your duplicate marked BAM file</strong></p>
<ul>
<li><p>In the <code>src</code> directory create a new script to configure Strelka2</p>
<ul>
<li>Ensure your variables are set appropriately and that you use the <code>\</code> operator to seperate commands over lines.</li>
</ul></li>
<li><p>Save your script and execute it</p></li>
</ul>
</blockquote>
</section>
<section id="running-the-strelka2-workflow" class="level2" data-number="9.4">
<h2 data-number="9.4" class="anchored" data-anchor-id="running-the-strelka2-workflow"><span class="header-section-number">9.4</span> Running the Strelka2 workflow</h2>
<p>We can run <code>Strelka2</code> by executing the <code>runWorkflow.py</code> script that we have generated. This script will be located in the directory you specify in the <code>--runDir</code> option in the configuration script.</p>
<blockquote class="blockquote">
<p><strong>Exercise 2: Run Strelka2 using the runWorkflow.py script</strong></p>
<ul>
<li><p>Navigate to the directory you set in the <code>--runDir</code> option of your configuration script.</p></li>
<li><p>In the terminal type <code>python2 runWorkflow.py -m local -j 6</code></p></li>
<li><p>This will run Strelka2.</p>
<ul>
<li><p>The <code>-m</code> option sets the executor to run locally on the VM.</p></li>
<li><p>The <code>-j</code> option instructs Strelka2 to use 6 cores.</p></li>
</ul></li>
</ul>
</blockquote>
</section>
<section id="evaluating-strelka2-variant-calling" class="level2" data-number="9.5">
<h2 data-number="9.5" class="anchored" data-anchor-id="evaluating-strelka2-variant-calling"><span class="header-section-number">9.5</span> Evaluating Strelka2 Variant calling</h2>
<p>Strelka2 will produce a number of files in a <code>results</code> sub-directory. The most important file here is the <code>variants.vcf.gz</code> file. This contains variant calls from your alignment in the form of variant call file (VCF4.1, <a href="http://samtools.github.io/hts-specs/VCFv4.1.pdf">click here for more details on this file format</a>). In week 2, we used standard bash commands to explore VCF files. Let’s take a closer look at our VCF file.</p>
<blockquote class="blockquote">
<p><strong>Exercise 3: Calculate the number of raw variant calls in your variants.vcf.gz file</strong></p>
<ul>
<li>Use <code>zcat</code>, <code>grep</code> and <code>wc</code> to evaluate the number of raw variant calls present in your VCF file.</li>
</ul>
</blockquote>
<p>Information regarding variant filtering is contained in the FILTER column. If the variant passed all filters this will be indicated by the string <code>PASS</code>. We can use this knowledge to identify the number of variants passing all filters.</p>
<blockquote class="blockquote">
<p><strong>Exercise 4: Calculate the number of variant calls that passed automated filtering</strong></p>
<ul>
<li>Use <code>zcat</code>, <code>grep</code> and <code>wc</code> to evaluate the number of variants with the string <code>PASS</code> in the filter column.</li>
</ul>
</blockquote>
<p>We can perform similar tasks using <code>bcftools</code>. <code>bcftools</code> is a bespoke tool that is designed specifically for analysing and manipulating VCF files. <a href="https://samtools.github.io/bcftools/bcftools.html">Click here to review the documentation for <code>bcftools</code>.</a> You will notice that there are several helpful features included in this tool. The ones we will focus on is the <code>bcftools view</code> and <code>bcftools stats</code> functions.</p>
<p><code>bcftools view</code> allows you to filter and manipulate the <code>vcf</code> file. You can filter the file by chromosome position, filter criteria, and other aspects of the file. Here is an example of how we might use <code>bcftools view</code> to subset a VCF file to include only variants that passed filtration. This example also calculates statistics on the variants passing filtering.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="va">VARIANT_DIRECTORY</span><span class="op">=</span>/home/trainee42/GNA5012/week2_5/results/variant_calling</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="va">SAMPLE</span><span class="op">=</span>SAMPLE_NAME</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-f</span> PASS <span class="va">${VARIANT_DIRECTORY}</span>/<span class="va">${SAMPLE}</span>/results/variants/variants.vcf.gz <span class="op">&gt;</span> <span class="va">${VARIANT_DIRECTORY}</span>/<span class="va">${SAMPLE}</span>/results/variants/<span class="va">${SAMPLE}</span>_PASS.vcf</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats <span class="va">${VARIANT_DIRECTORY}</span>/<span class="va">${SAMPLE}</span>/results/variants/<span class="va">${SAMPLE}</span>_PASS.vcf <span class="op">&gt;</span>  <span class="va">${VARIANT_DIRECTORY}</span>/<span class="va">${SAMPLE}</span>/results/variants/<span class="va">${SAMPLE}</span>_PASS_Stats.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p><strong>Exercise 5: Use bcftools to subset your VCF file to include only variants passing filtering.</strong></p>
<ul>
<li>In the <code>src</code> directory create a new script to subset variants and calculate statistics on the number of variants passing filters.</li>
<li>Open the statistics and evaluate the number of SNPs and INDELs</li>
<li>Calculate statistics on the unfiltered VCF file. What do you notice?</li>
</ul>
</blockquote>
</section>
<section id="empirical-variant-scoring-filtration-in-strelka2" class="level2" data-number="9.6">
<h2 data-number="9.6" class="anchored" data-anchor-id="empirical-variant-scoring-filtration-in-strelka2"><span class="header-section-number">9.6</span> Empirical variant scoring filtration in Strelka2</h2>
<p>Strelka2 uses pretrained random forest machine learning models to estimate the liklihood that a variant is erroneous and due to sequencing error. Each variant is scored against this model for several criterion. If a variant achieves a sufficiently high score it will fail that filter. <a href="https://www.biorxiv.org/content/biorxiv/suppl/2017/09/25/192872.DC2/192872-1.pdf">Click here for more details on the EVS model method employed by Strelka2.</a></p>
<blockquote class="blockquote">
<p><strong>Exercise 6: Create a summary of the germline filtration criteria considered by Strelka2. Discuss them at your table.</strong></p>
</blockquote>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation column-page-right">
  <div class="nav-page nav-page-previous">
      <a href="./W4_1.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Week 4 Workshop A - Post-Alignment Processing</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->



</body></html>