<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>GNA5012 - Applied Bioinformatics - 10&nbsp; Week 5 Workshop A - Variant Filtration and visualising variant calls</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./W6_1.html" rel="next">
<link href="./W4_2.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/videojs/video.min.js"></script>
<link href="site_libs/quarto-contrib/videojs/video-js.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Week 5 Workshop A - Variant Filtration and visualising variant calls</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">GNA5012 - Applied Bioinformatics</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assessment.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Assessment 1 - Capstone A</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W1_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Week 1 Workshop A - Introduction to the shell</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W1_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Week 1 Workshop B - Introduction shell scripting and pipelines</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W2_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Week 2 Workshop A - Bioinformatic File Types</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W2_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Week 2 Workshop B - Quality control inspection of sequencing reads</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W3_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Week 3 Workshop A - Quality control and read trimming</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W3_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Week 3 Workshop B - Sequence Alignment</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W4_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Week 4 Workshop A - Post-Alignment Processing</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W4_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Week 4 Workshop B - Variant Calling</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W5_1.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Week 5 Workshop A - Variant Filtration and visualising variant calls</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W6_1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Week 6 Workshop A - A brief introduction to R in bioinformatics analyses</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W6_2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Week 6 Workshop B - Reproducible research and R notebooks</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./W8a.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Week 8 Workshop A - Basic R data structures and variable types</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content column-page-right" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Week 5 Workshop A - Variant Filtration and visualising variant calls</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>
<nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives"><span class="toc-section-number">10.1</span>  Learning Objectives</a></li>
  <li><a href="#workshop-setup" id="toc-workshop-setup"><span class="toc-section-number">10.2</span>  Workshop Setup</a></li>
  <li><a href="#hard-filtering-variants" id="toc-hard-filtering-variants"><span class="toc-section-number">10.3</span>  Hard filtering variants</a></li>
  <li><a href="#visualising-sequence-alignments" id="toc-visualising-sequence-alignments"><span class="toc-section-number">10.4</span>  Visualising sequence alignments</a></li>
  </ul>
</nav>
<p>This chapter represents the final step in the variant calling process. Last week we generated both a raw and EVS-filtered variant call file for our genome-in-a-bottle exome sequencing experiment. Today we will explore additional filtration and visualise our variants to inspect what a “good” and “bad” variant call looks like. We will be using a tool called <a href="https://software.broadinstitute.org/software/igv/download">IGV</a>. IGV is a genome browser and can be used to load in sequence alignments and visualise them. We can use IGV to manually inspect variants. By manually inspecting variants that pass filtration and fail filtration we can begin to understand the variant filtration process and its importance.</p>
<div class="cell">

</div>
<section id="learning-objectives" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="learning-objectives"><span class="header-section-number">10.1</span> Learning Objectives</h2>
<p>At the conclusion of today’s workshop students are expected to be able to:</p>
<ul>
<li><p>Perform additional hard filtering of variants using <code>bcftools</code></p></li>
<li><p>Visualise sequence alignments using <code>IGV</code> and examine variant calls</p></li>
<li><p>Explain common reasons for variants failing filtration, with reference to the specific reason a variant has been filtered.</p></li>
</ul>
</section>
<section id="workshop-setup" class="level2" data-number="10.2">
<h2 data-number="10.2" class="anchored" data-anchor-id="workshop-setup"><span class="header-section-number">10.2</span> Workshop Setup</h2>
<p>To prepare for today’s workshop activities you need to</p>
<ul>
<li><p>Download <a href="https://software.broadinstitute.org/software/igv/download">IGV</a>. IGV is a java application and is compatible with Windows and Mac OSX based computers. We will be loading in <code>BAM</code> and <code>BAI</code> files that will consume ~3GB of RAM. If you run into issues regarding memory consumption please get the attention of one of the teaching staff who will propose alternatives.</p></li>
<li><p>Download the duplicate marked <code>BAM</code> and <code>BAI</code> files you generated in Week Four to your laptop. Also download your <code>variants.vcf.gz</code> file from the variants results directory. Place all of these files in the same directory on your local PC</p></li>
</ul>
</section>
<section id="hard-filtering-variants" class="level2" data-number="10.3">
<h2 data-number="10.3" class="anchored" data-anchor-id="hard-filtering-variants"><span class="header-section-number">10.3</span> Hard filtering variants</h2>
<p>Our VCF file we generated last week (<code>variants.vcf.gz</code>) has already been filtered using EVS, however we can explore how well hard-filtration matches the EVS-based filtration using <code>bcftools</code>. You can use <code>bcftools</code> to filter based on any metric in the <code>FORMAT</code> column of the <code>.vcf</code> file. You can find the metrics available in the <code>FORMAT</code> column and their description in the header of the VCF file.</p>
<blockquote class="blockquote">
<p><strong>Exercise 1: Using zcat, grep and less, view the header of the VCF file and the options available in the FORMAT column</strong></p>
<ul>
<li><p>You can enter this command directly into the terminal, however take note of how you do it for future reference.</p></li>
<li><p>You will need to use pipes to complete this task.</p></li>
<li><p>Discuss at your tables the metrics generated by Strelka2 in the <code>FORMAT</code> column of the VCF files. What information can you glean from them?</p></li>
</ul>
</blockquote>
<p>Now that you’ve identified what metrics are available to you, you can use <code>bcftools</code> to perform hard-filtering on them. The standard syntax of <code>bcftools</code> for filtration is:</p>
<ul>
<li><code>bcftools view -i 'METRIC &gt;= VALUE' input_vcf.vcf.gz</code></li>
</ul>
<p>Let’s break this down</p>
<ul>
<li><p><code>bcftools view</code> - This will open the <code>vcf</code> file and return it to standard output</p></li>
<li><p><code>-i</code> this option says to include variants that meet a given criteria</p></li>
<li><p><code>'METRIC &gt;= VALUE'</code> - Here we replace <code>METRIC</code> with an expression that is present in the <code>FORMAT</code> column and <code>VALUE</code> with our desired threshold. For example</p>
<ul>
<li><code>'DP &gt;= 20'</code> - This would only include variants with a depth of &gt; 20 reads.</li>
</ul></li>
</ul>
<blockquote class="blockquote">
<p><strong>Exercise 2: Use <code>bcftools</code></strong> <strong>to filter variants based on a depth threshold of 10</strong></p>
<ul>
<li><p>In your src directory, create a new script called <code>depth_filter.sh</code></p>
<ul>
<li><p>Set up variables to point to your vcf file. Use the principles we learned in previous weeks to set your variables appropriately</p></li>
<li><p>Use DP &gt;=10 as the filter expression</p></li>
<li><p>Redirect your output to a new vcf file called <code>variants_DP_filtered.vcf</code> in the results directory for your variants.</p></li>
</ul></li>
<li><p>How many variants were filtered out using a simple depth threshold of 10?</p></li>
</ul>
</blockquote>
<p>We can string filter criteria together using the <code>&amp;&amp;</code> and <code>||</code> operators. These refer to AND and OR, respectively. If you use the <code>&amp;&amp;</code> operator, variants must meet all criteria to be included. If you use the <code>||</code> operator, if a variant meets one of the criteria it will be included in the output. For example:</p>
<ul>
<li><p><code>bcftools view -i 'DP &gt;= 20 &amp;&amp; GQX &gt;= 30'</code> : This expression will include only variants with a depth of &gt;= 20 <strong>and</strong> a genotype quality of &gt;=30.</p></li>
<li><p><code>bcftools view -i 'DP &gt;= 20 || GQX &gt;= 30'</code> : This expression will include variants with a depth of &gt;=20 <strong>or</strong> a genotype quality of &gt;=30.</p></li>
</ul>
<blockquote class="blockquote">
<p><strong>Exercise 3: Use <code>bcftools</code></strong> <strong>to filter variants based on a depth threshold of 10 AND/OR a GQX of &gt;=20</strong></p>
<ul>
<li><p>In your src directory, create a new script called <code>depth_GQX_filter.sh</code></p>
<ul>
<li><p>Set up variables to point to your vcf file. Use the principles we learned in previous weeks to set your variables appropriately</p></li>
<li><p>Write two lines of code using <code>bcftools</code>. The first should filter using the <code>&amp;&amp;</code> operator, redirecting the output to a new VCF file. The second should filter using the <code>||</code> operator, redirecting the output to a different VCF file.</p></li>
<li><p>Remember to name the output files appropriately.</p></li>
</ul></li>
<li><p>Compare and contrast the resulting VCF files. What do you notice about the number of variants in each? Which is more permissive?</p></li>
</ul>
</blockquote>
</section>
<section id="visualising-sequence-alignments" class="level2" data-number="10.4">
<h2 data-number="10.4" class="anchored" data-anchor-id="visualising-sequence-alignments"><span class="header-section-number">10.4</span> Visualising sequence alignments</h2>
<p>Sequence alignments and variant calls can be visualised using interactive tools such as <code>IGV</code>. <code>IGV</code> is a genome browser that will take as input a <code>BAM</code> file. <code>IGV</code> will only work if an accompanying <code>.BAI</code> index file is found in the same directory as the <code>BAM</code> file. Watch the following short videos on loading <code>BAM</code> files into <code>IGV</code>, sequencing data basics and inspecting SNPs in <code>IGV</code>.</p>
<p></p><div class="quarto-video"><video id="video_shortcode_videojs_video1" width="250" height="175" class="video-js vjs-default-skin " controls="" preload="auto" data-setup="{}" title=""><source src="images/IGV.mp4"></video></div> <div class="quarto-video"><iframe src="https://www.youtube.com/embed/E_G8z_2gTYM" width="250" height="175" title="" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen=""></iframe></div> <div class="quarto-video"><iframe src="https://www.youtube.com/embed/eZUI6lcIX3I" width="250" height="175" title="" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen=""></iframe></div><p></p>
<p>There are a number of reasons why variants are filtered by Strelka2. These are outlined in the weeks readings and also in the header of the VCF file</p>
<p><img src="images/Screenshot 2023-08-22 at 8.21.04 pm.png" class="img-fluid"></p>
<blockquote class="blockquote">
<p><strong>Exercise 4: Using IGV, visualise variants that pass filtration and those that are filtered for each reason.</strong></p>
<ul>
<li><p>You will need to ensure that you load the <code>.BAM</code> and <code>.VCF</code> into IGV</p></li>
<li><p>Use grep to identify variants that failed filtration for each filtering criteria.</p></li>
<li><p>Take note of what the appearance of the variant and the alignment. Compare this with a variant that passed filtration</p></li>
<li><p>Discuss this at your tables before we discuss as a group.</p></li>
</ul>
</blockquote>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation column-page-right">
  <div class="nav-page nav-page-previous">
      <a href="./W4_2.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Week 4 Workshop B - Variant Calling</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./W6_1.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Week 6 Workshop A - A brief introduction to R in bioinformatics analyses</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<script>videojs(video_shortcode_videojs_video1);</script>



</body></html>